pkg/packagekit/updates.jsx

```diff
@@ -442,21 +447,29 @@ class OsUpdates extends React.Component {
     }
 
     watchUpdates(transProxy) {
-        this.setState({state: "applying", applyTransaction: transProxy});
+        this.setState({state: "applying", applyTransaction: transProxy, allowCancel: transProxy.AllowCancel});
 
         transProxy.addEventListener("ErrorCode", (event, code, details) => this.state.errorMessages.push(details));
+
         transProxy.addEventListener("Finished", (event, exit) => {
-            if (exit == PK_EXIT_ENUM_SUCCESS) {
+            this.setState({applyTransaction: null, allowCancel: null});
+
+            if (exit == PK_EXIT_ENUM_SUCCESS || exit == PK_EXIT_ENUM_CANCELLED) {
                 this.setState({state: "loading", haveSecurity: false, loadPercent: null});
                 this.loadUpdates();
             } else {
                 // normally we get FAILED here with ErrorCodes; handle unexpected errors to allow for some debugging
                 if (exit != PK_EXIT_ENUM_FAILED)
-                    this.state.errorMessages.push(cockpit.format(_("PackageKit reported error code {0}"), exit));
+                    this.state.errorMessages.push(cockpit.format(_("PackageKit reported error code $0"), exit));
                 this.setState({state: "updateError"});
             }
         });
 
+        transProxy.addEventListener("changed", (event, data) => {
+            if ("AllowCancel" in data && data.AllowCancel != this.state.allowCancel)
```

Comment:
As far as I understood it will always recompute the virtual DOM (i. e. re-evaluate all the component `render()` functions) and then just compare with the current real DOM. I'm fairly sure that `setState()` does not automatically reduce to a no-op if you set the same value again, as you can do something like
```js
this.state.foo = 2;
this.setState({foo: this.state.foo});
```
and it will recompute the tree.

I did that because there are a *lot* of `changed` events, several per second. However, I may be overly cautious, so if it's general best practice to not worry about this, I'm more than happy to drop this.